### PR TITLE
Add a way to extract function spaces as json files

### PR DESCRIPTION
**Pull request by Luni-4**
This PR adds a way to extract function spaces as `json` files. It can work both on a single file and on a directory.

Thanks in advance for your review! :)

### CODE REVIEW

**Comment by calixteman** [comment]
This patch looks pretty good :) but there is already an option "-o" to set the output so I'd say that "--metrics -o foo" should dump the json in foo.
We should probably dump on stdout or in a file according to "-o" presence:
https://github.com/mozilla/rust-code-analysis/blob/master/src/metrics.rs#L396

**Comment by Luni-4** [comment]
> This patch looks pretty good :) but there is already an option "-o" to set the output so I'd say that "--metrics -o foo" should dump the json in foo.

I tried to run the following command: `./rust-code-analysis --metrics -o foo -p file.rs`

But it doesn't create the `foo` file since the file creation is related to the preproc part as you can see from [here](https://github.com/mozilla/rust-code-analysis/blob/master/src/main.rs#L531)

> We should probably dump on stdout or in a file according to "-o" presence:
> https://github.com/mozilla/rust-code-analysis/blob/master/src/metrics.rs#L396

Good idea! We could add a `json_bool` field to the `MetricsCfg` struct and move `dump_json` into `FuncSpace`. But what about the `-o` option? It works differently as it is now.

**Comment by calixteman** [comment]
"-o" isn't especially related to "-preproc".
What I proposed to you is just to modify your patch to get the "-o" value, pass it to Config and when used with metrics=true, then pass it into MetricsCfg... and then if we've an output path in MetricsCfg then dump the json in it else dump on stdout.

**Comment by Luni-4** [comment]
> "-o" isn't especially related to "-preproc".

I see, thanks for your explanation!

> What I proposed to you is just to modify your patch to get the "-o" value, pass it to Config and when used with metrics=true, then pass it into MetricsCfg... and then if we've an output path in MetricsCfg then dump the json in it else dump on stdout.

Yep, I'm going to do it, thanks!

File: src/main.rs [hunk 1]
```diff
@@ -72,8 +73,20 @@ fn act_on_file(language: LANG, path: PathBuf, cfg: Config) -> std::io::Result<()
     } else if cfg.metrics {
         let source = read_file_with_eol(&path)?;
         let language = guess_language(&source).unwrap_or(language);
-        let cfg = MetricsCfg { path };
-        action::<Metrics>(&language, source, &cfg.path.clone(), pr, cfg)
+        if cfg.output.is_empty() {
```
Comment (calixteman) [inline src/main.rs]: nit: there is some dup code in then & else clauses (finally the only difference is almost output_path definition in the struct)

File: src/main.rs [hunk 2]
```diff
@@ -320,7 +333,7 @@ fn main() {
         )
         .arg(
             Arg::with_name("output")
-                .help("Output file")
+                .help("Output directory")
```
Comment (calixteman) [inline src/main.rs]: I'd say `file/directory`.
`file` was too restrictive but `directory` is too

File: src/main.rs [hunk 3]
```diff
@@ -439,7 +452,12 @@ fn main() {
         (None, None)
     };
 
-    let output = matches.value_of("output").unwrap();
+    let output = matches.value_of("output").unwrap().to_string();
+    let output_is_dir = PathBuf::from(output.clone()).is_dir();
+    if !output.is_empty() && !output_is_dir {
```
Comment (calixteman) [inline src/main.rs]: We can adapt according to the kind of path we've and the tool we're using: I mean `--metrics -o not_dir` is forbidden but can be allowed with `--foo -o not_dir`.
So just disallow file type when we're running metrics.

File: src/metrics.rs [hunk 4]
```diff
@@ -306,6 +307,17 @@ impl<'a> FuncSpace<'a> {
         color!(stdout, White);
         writeln!(stdout, "{}", val)
     }
+
+    fn dump_json(&self, path: &PathBuf, output_path: &PathBuf) -> std::io::Result<()> {
+        let json_data = serde_json::to_string_pretty(&self).unwrap();
```
Comment (calixteman) [inline src/metrics.rs]: Could you file a bug to add an option `--pretty` to choose if the dump should be pretty or not ?

File: src/main.rs (hunk 3 above)
Comment (Luni-4) [reply src/main.rs]: You're right, I will change that, thanks!

File: src/metrics.rs (hunk 4 above)
Comment (Luni-4) [reply src/metrics.rs]: Sure! :)

Comment (Luni-4) [reply src/metrics.rs]: Done

File: src/main.rs (hunk 1 above)
Comment (Luni-4) [reply src/main.rs]: Fixed

File: src/metrics.rs [hunk 5]
```diff
@@ -306,6 +307,17 @@ impl<'a> FuncSpace<'a> {
...
+
+        let mut file = path.as_path().file_name().unwrap().to_os_string();
+        file.push(".json");
+
+        let mut json_path = output_path.clone();
+        json_path.push(file);
+        write_file(&json_path, json_data.to_string().as_bytes())
```
Comment (calixteman) [inline src/metrics.rs]: nit: the `to_string` is probably useless

Comment (calixteman) [inline src/metrics.rs]: how to deal with an existing file ?
For example, you have `./foo/a.py` and `./bar/a.py`, the two files `a.py.json` will be the same.

Comment (Luni-4) [reply src/metrics.rs]: We could look if there is already an `a.py.json` file in the `output_path` directory and replace filenames with paths (e.g `./foo/a.py.json` and ./bar/a.py.json`) 

What do you think of this approach?

Comment (calixteman) [reply src/metrics.rs]: It's ok for me, let's do that and if someone complains at some point we'll add an option or whatever...

Comment (Luni-4) [reply src/metrics.rs]: > It's ok for me, let's do that and if someone complains at some point we'll add an option or whatever...

I agree

> FYI, @mozilla we use the server part for now.

I see, that's why is full of tests, thanks for the info! :)

Comment (Luni-4) [reply src/metrics.rs]: @calixteman 
Done. I've replaced slashes, backslashes and colons with `_`.

**Comment by Luni-4** [comment]
@calixteman 
Is still there something I need to do such that this PR could be landed?

**Comment by calixteman** [comment]
Grazie mille

**Comment by Luni-4** [comment]
Grazie a te :)

**Pull request merged**